## Import Libraries and Data

In [339]:
import pandas as pd
import numpy as np

In [340]:
# This code was given by Simon Clarke!!
# sep=',|\t' is another way to split the clusted data that were not splited by ',' as they should be (csv = comma-seperated values)
# on_bad_lines='warn' is just to signal a warning that there are bad lines in the data (it will just 'warn' us, not 'skip' like what we did)

raw_df = pd.read_csv('Data/Melbourne01.csv',names=["Year", "Month", "Day", "Hour", "Minute",
                                          "Air Temp (degrees C)",
                                          "Apparent Temp (degrees C)", "Dew Pt Temp (degrees C)", "Humidity (%)",
                                          "Wind Direction", "Wind Speed (km/h)", "Wind Gust  (km/h)",
                                          "MSLP (hPa)",
                                          "Rainfall since 9 am (mm)", "Dummy1", "Dummy2"],
                sep=',|\t', on_bad_lines='warn')

C:\Users\Zenbook\AppData\Local\Temp\ipykernel_1984\1274739746.py:5: ParserWarning: Falling back to the 'python' engine because the 'c' engine does not support regex separators (separators > 1 char and different from '\s+' are interpreted as regex); you can avoid this warning by specifying engine='python'.
  raw_df = pd.read_csv('Data/Melbourne01.csv',names=["Year", "Month", "Day", "Hour", "Minute",


In [341]:
raw_df

# Just to get an idea what the raw data looks like

,Year,Month,Day,Hour,Minute,Air Temp (degrees C),Apparent Temp (degrees C),Dew Pt Temp (degrees C),Humidity (%),Wind Direction,Wind Speed (km/h),Wind Gust (km/h),MSLP (hPa),Rainfall since 9 am (mm),Dummy1,Dummy2
0,2011.0,1,1,0,4,24.8,0.0,14.0,51.0,SE,11,13.0,1007.4,0.0,NaN,NaN
1,2011.0,1,1,0,4,24.8,0.0,14.0,51.0,SE,11,13.0,1007.4,0.0,NaN,NaN
2,2011.0,1,1,0,14,24.8,0.0,13.3,48.0,SE,11,11.0,1007.5,0.0,NaN,NaN
3,2011.0,1,1,0,14,24.8,0.0,13.3,48.0,SE,11,11.0,1007.5,0.0,NaN,NaN
4,2011.0,1,1,0,24,24.9,0.0,13.3,48.0,SE,11,13.0,1007.5,0.0,NaN,NaN
...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...
1113280,2025.0,3,12,15,30,33.4,32.5,14.9,32.0,NW,20.0,35.0,1014.6,0.0,NaN,NaN
1113281,2025.0,3,12,15,34,33.5,32.9,14.5,31.0,NNW,26.0,41.0,1014.6,0.0,NaN,NaN
1113282,2025.0,3,12,16,0,34.4,34.2,14.8,30.0,WNW,26.0,41.0,1014.4,0.0,NaN,NaN
1113283,2025.0,3,12,16,30,33.6,32.4,14.1,30.0,WNW,26.0,41.0,1014.1,0.0,NaN,NaN


In [342]:
print(raw_df.isna().sum())

# 133 rows of missing value in 'Year' column is due to the shfted thing...
# For Dummy1, Dummy2 we have no idea what they are....

Year                             133
Month                              0
Day                                0
Hour                               0
Minute                             0
Air Temp (degrees C)               0
Apparent Temp (degrees C)          0
Dew Pt Temp (degrees C)            0
Humidity (%)                       0
Wind Direction                     0
Wind Speed (km/h)                  0
Wind Gust  (km/h)                  0
MSLP (hPa)                         0
Rainfall since 9 am (mm)           0
Dummy1                       1113152
Dummy2                       1113152
dtype: int64


## Data Formating

In [343]:
df = raw_df.drop_duplicates()    # drop duplicated rows
df = df.reset_index(drop=True)   # resetting the index after dropping some rows
df.tail()

,Year,Month,Day,Hour,Minute,Air Temp (degrees C),Apparent Temp (degrees C),Dew Pt Temp (degrees C),Humidity (%),Wind Direction,Wind Speed (km/h),Wind Gust (km/h),MSLP (hPa),Rainfall since 9 am (mm),Dummy1,Dummy2
576083,2025.0,3,12,15,30,33.4,32.5,14.9,32.0,NW,20.0,35.0,1014.6,0.0,NaN,NaN
576084,2025.0,3,12,15,34,33.5,32.9,14.5,31.0,NNW,26.0,41.0,1014.6,0.0,NaN,NaN
576085,2025.0,3,12,16,0,34.4,34.2,14.8,30.0,WNW,26.0,41.0,1014.4,0.0,NaN,NaN
576086,2025.0,3,12,16,30,33.6,32.4,14.1,30.0,WNW,26.0,41.0,1014.1,0.0,NaN,NaN
576087,2025.0,3,12,16,30,33.6,32.4,14.1,30.0,NW,28.0,35.0,1014.1,0.0,NaN,NaN


In [344]:
shifted_rows_index = df.index[df['Year'].isna()].tolist()  # list out all the shifted rows (missing value in 'Year' column)
df.loc[shifted_rows_index]

,Year,Month,Day,Hour,Minute,Air Temp (degrees C),Apparent Temp (degrees C),Dew Pt Temp (degrees C),Humidity (%),Wind Direction,Wind Speed (km/h),Wind Gust (km/h),MSLP (hPa),Rainfall since 9 am (mm),Dummy1,Dummy2
121011,NaN,2013,9,6,17,0.0,14.9,11.4,5.6,53,N,20.0,22.0,1024.3,0.0,53.0
121012,NaN,2013,9,6,17,30.0,14.7,11.8,6.4,57,N,13.0,17.0,1024.2,0.0,57.0
121013,NaN,2013,9,6,18,0.0,14.4,11.6,7.0,61,N,13.0,17.0,1024.3,0.0,61.0
121014,NaN,2013,9,6,18,30.0,14.1,11.1,7.6,65,N,17.0,20.0,1024.1,0.0,65.0
121015,NaN,2013,9,6,19,0.0,14.1,10.6,7.1,62,N,15.0,19.0,1024.5,0.0,62.0
...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...
121139,NaN,2013,9,9,9,0.0,16.3,6.5,2.0,38,N,56.0,70.0,1012,0.2,38.0
121140,NaN,2013,9,9,9,30.0,17.1,6.2,1.3,34,N,56.0,72.0,1011.4,0.0,34.0
121141,NaN,2013,9,9,10,0.0,18.0,7.6,-0.2,29,N,56.0,74.0,1010.8,0.0,29.0
121142,NaN,2013,9,9,10,30.0,18.6,7.7,-1.0,26,N,54.0,69.0,1010.2,0.0,26.0


From the table above, we can see that ther are no values in the Year column because all the data are shifted one column to the right.

### Dealing with shifted rows

In [345]:
df_fixing = df.loc[shifted_rows_index]  # Create another df for those shifted rows
df_fixing

,Year,Month,Day,Hour,Minute,Air Temp (degrees C),Apparent Temp (degrees C),Dew Pt Temp (degrees C),Humidity (%),Wind Direction,Wind Speed (km/h),Wind Gust (km/h),MSLP (hPa),Rainfall since 9 am (mm),Dummy1,Dummy2
121011,NaN,2013,9,6,17,0.0,14.9,11.4,5.6,53,N,20.0,22.0,1024.3,0.0,53.0
121012,NaN,2013,9,6,17,30.0,14.7,11.8,6.4,57,N,13.0,17.0,1024.2,0.0,57.0
121013,NaN,2013,9,6,18,0.0,14.4,11.6,7.0,61,N,13.0,17.0,1024.3,0.0,61.0
121014,NaN,2013,9,6,18,30.0,14.1,11.1,7.6,65,N,17.0,20.0,1024.1,0.0,65.0
121015,NaN,2013,9,6,19,0.0,14.1,10.6,7.1,62,N,15.0,19.0,1024.5,0.0,62.0
...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...
121139,NaN,2013,9,9,9,0.0,16.3,6.5,2.0,38,N,56.0,70.0,1012,0.2,38.0
121140,NaN,2013,9,9,9,30.0,17.1,6.2,1.3,34,N,56.0,72.0,1011.4,0.0,34.0
121141,NaN,2013,9,9,10,0.0,18.0,7.6,-0.2,29,N,56.0,74.0,1010.8,0.0,29.0
121142,NaN,2013,9,9,10,30.0,18.6,7.7,-1.0,26,N,54.0,69.0,1010.2,0.0,26.0


In [346]:
df_fixing.drop(columns=['Year'], inplace=True)  # Droping Year column since it has no data in it

In [347]:
col_names=['Year', "Month", "Day", "Hour", "Minute", "Air Temp (degrees C)", "Apparent Temp (degrees C)", 
           "Dew Pt Temp (degrees C)", "Humidity (%)", "Wind Direction", "Wind Speed (km/h)", "Wind Gust  (km/h)",
            "MSLP (hPa)", "Rainfall since 9 am (mm)", "Dummy1"]

df_fixing.columns = col_names # Rename all the column names after dropping Year column
df_fixing

,Year,Month,Day,Hour,Minute,Air Temp (degrees C),Apparent Temp (degrees C),Dew Pt Temp (degrees C),Humidity (%),Wind Direction,Wind Speed (km/h),Wind Gust (km/h),MSLP (hPa),Rainfall since 9 am (mm),Dummy1
121011,2013,9,6,17,0.0,14.9,11.4,5.6,53,N,20.0,22.0,1024.3,0.0,53.0
121012,2013,9,6,17,30.0,14.7,11.8,6.4,57,N,13.0,17.0,1024.2,0.0,57.0
121013,2013,9,6,18,0.0,14.4,11.6,7.0,61,N,13.0,17.0,1024.3,0.0,61.0
121014,2013,9,6,18,30.0,14.1,11.1,7.6,65,N,17.0,20.0,1024.1,0.0,65.0
121015,2013,9,6,19,0.0,14.1,10.6,7.1,62,N,15.0,19.0,1024.5,0.0,62.0
...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...
121139,2013,9,9,9,0.0,16.3,6.5,2.0,38,N,56.0,70.0,1012,0.2,38.0
121140,2013,9,9,9,30.0,17.1,6.2,1.3,34,N,56.0,72.0,1011.4,0.0,34.0
121141,2013,9,9,10,0.0,18.0,7.6,-0.2,29,N,56.0,74.0,1010.8,0.0,29.0
121142,2013,9,9,10,30.0,18.6,7.7,-1.0,26,N,54.0,69.0,1010.2,0.0,26.0


We can see that all the data are shifted back and have correct column names. But you can see that 'Dummy2' column is gone. But we don't really care (right?) because we don't know what it is for.

### Merge them back together

In [348]:
df.drop(df.index[shifted_rows_index], inplace=True) # Drop all the shifted rows from 'df' to get only non-shifted rows
df

,Year,Month,Day,Hour,Minute,Air Temp (degrees C),Apparent Temp (degrees C),Dew Pt Temp (degrees C),Humidity (%),Wind Direction,Wind Speed (km/h),Wind Gust (km/h),MSLP (hPa),Rainfall since 9 am (mm),Dummy1,Dummy2
0,2011.0,1,1,0,4,24.8,0.0,14.0,51.0,SE,11,13.0,1007.4,0.0,NaN,NaN
1,2011.0,1,1,0,14,24.8,0.0,13.3,48.0,SE,11,11.0,1007.5,0.0,NaN,NaN
2,2011.0,1,1,0,24,24.9,0.0,13.3,48.0,SE,11,13.0,1007.5,0.0,NaN,NaN
3,2011.0,1,1,0,34,24.7,0.0,13.4,49.0,SE,11,11.0,1007.4,0.0,NaN,NaN
4,2011.0,1,1,0,44,24.1,0.0,13.3,51.0,ESE,9,9.0,1007.3,0.0,NaN,NaN
...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...
576083,2025.0,3,12,15,30,33.4,32.5,14.9,32.0,NW,20.0,35.0,1014.6,0.0,NaN,NaN
576084,2025.0,3,12,15,34,33.5,32.9,14.5,31.0,NNW,26.0,41.0,1014.6,0.0,NaN,NaN
576085,2025.0,3,12,16,0,34.4,34.2,14.8,30.0,WNW,26.0,41.0,1014.4,0.0,NaN,NaN
576086,2025.0,3,12,16,30,33.6,32.4,14.1,30.0,WNW,26.0,41.0,1014.1,0.0,NaN,NaN


In [349]:
all_df = pd.concat([df, df_fixing])  # Combining the non-shifted df with the shifted df (which has already been fixed)
all_df = all_df.sort_index()         # Sort all those rows back into place
all_df

,Year,Month,Day,Hour,Minute,Air Temp (degrees C),Apparent Temp (degrees C),Dew Pt Temp (degrees C),Humidity (%),Wind Direction,Wind Speed (km/h),Wind Gust (km/h),MSLP (hPa),Rainfall since 9 am (mm),Dummy1,Dummy2
0,2011.0,1,1,0,4.0,24.8,0.0,14.0,51.0,SE,11,13.0,1007.4,0.0,NaN,NaN
1,2011.0,1,1,0,14.0,24.8,0.0,13.3,48.0,SE,11,11.0,1007.5,0.0,NaN,NaN
2,2011.0,1,1,0,24.0,24.9,0.0,13.3,48.0,SE,11,13.0,1007.5,0.0,NaN,NaN
3,2011.0,1,1,0,34.0,24.7,0.0,13.4,49.0,SE,11,11.0,1007.4,0.0,NaN,NaN
4,2011.0,1,1,0,44.0,24.1,0.0,13.3,51.0,ESE,9,9.0,1007.3,0.0,NaN,NaN
...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...
576083,2025.0,3,12,15,30.0,33.4,32.5,14.9,32.0,NW,20.0,35.0,1014.6,0.0,NaN,NaN
576084,2025.0,3,12,15,34.0,33.5,32.9,14.5,31.0,NNW,26.0,41.0,1014.6,0.0,NaN,NaN
576085,2025.0,3,12,16,0.0,34.4,34.2,14.8,30.0,WNW,26.0,41.0,1014.4,0.0,NaN,NaN
576086,2025.0,3,12,16,30.0,33.6,32.4,14.1,30.0,WNW,26.0,41.0,1014.1,0.0,NaN,NaN


So, we now have a dataframe with all the data in the same format.

## Data Cleaning

In [350]:
# Dropping all the non-necessary columns
all_df.drop(columns=['Dummy1', 'Dummy2', 'Apparent Temp (degrees C)','Wind Gust  (km/h)'], inplace=True)
all_df

,Year,Month,Day,Hour,Minute,Air Temp (degrees C),Dew Pt Temp (degrees C),Humidity (%),Wind Direction,Wind Speed (km/h),MSLP (hPa),Rainfall since 9 am (mm)
0,2011.0,1,1,0,4.0,24.8,14.0,51.0,SE,11,1007.4,0.0
1,2011.0,1,1,0,14.0,24.8,13.3,48.0,SE,11,1007.5,0.0
2,2011.0,1,1,0,24.0,24.9,13.3,48.0,SE,11,1007.5,0.0
3,2011.0,1,1,0,34.0,24.7,13.4,49.0,SE,11,1007.4,0.0
4,2011.0,1,1,0,44.0,24.1,13.3,51.0,ESE,9,1007.3,0.0
...,...,...,...,...,...,...,...,...,...,...,...,...
576083,2025.0,3,12,15,30.0,33.4,14.9,32.0,NW,20.0,1014.6,0.0
576084,2025.0,3,12,15,34.0,33.5,14.5,31.0,NNW,26.0,1014.6,0.0
576085,2025.0,3,12,16,0.0,34.4,14.8,30.0,WNW,26.0,1014.4,0.0
576086,2025.0,3,12,16,30.0,33.6,14.1,30.0,WNW,26.0,1014.1,0.0


In [351]:
all_df.isna().sum() # Check for missing values

Year                        0
Month                       0
Day                         0
Hour                        0
Minute                      0
Air Temp (degrees C)        0
Dew Pt Temp (degrees C)     0
Humidity (%)                0
Wind Direction              0
Wind Speed (km/h)           0
MSLP (hPa)                  0
Rainfall since 9 am (mm)    0
dtype: int64

Since all the Year, Month, Day, Hour, Minute columns have no missing values, that means we can now create timestamp out of those columns.

### Timestamp

In [352]:
# Convert it into Timestamp format
all_df['Timestamp'] = pd.to_datetime(all_df[['Year','Month','Day','Hour','Minute']])
all_df.head()

,Year,Month,Day,Hour,Minute,Air Temp (degrees C),Dew Pt Temp (degrees C),Humidity (%),Wind Direction,Wind Speed (km/h),MSLP (hPa),Rainfall since 9 am (mm),Timestamp
0,2011.0,1,1,0,4.0,24.8,14.0,51.0,SE,11,1007.4,0.0,2011-01-01 00:04:00
1,2011.0,1,1,0,14.0,24.8,13.3,48.0,SE,11,1007.5,0.0,2011-01-01 00:14:00
2,2011.0,1,1,0,24.0,24.9,13.3,48.0,SE,11,1007.5,0.0,2011-01-01 00:24:00
3,2011.0,1,1,0,34.0,24.7,13.4,49.0,SE,11,1007.4,0.0,2011-01-01 00:34:00
4,2011.0,1,1,0,44.0,24.1,13.3,51.0,ESE,9,1007.3,0.0,2011-01-01 00:44:00


In [353]:
all_df.index = all_df['Timestamp']
all_df.head()

,Year,Month,Day,Hour,Minute,Air Temp (degrees C),Dew Pt Temp (degrees C),Humidity (%),Wind Direction,Wind Speed (km/h),MSLP (hPa),Rainfall since 9 am (mm),Timestamp
Timestamp,,,,,,,,,,,,,
2011-01-01 00:04:00,2011.0,1,1,0,4.0,24.8,14.0,51.0,SE,11,1007.4,0.0,2011-01-01 00:04:00
2011-01-01 00:14:00,2011.0,1,1,0,14.0,24.8,13.3,48.0,SE,11,1007.5,0.0,2011-01-01 00:14:00
2011-01-01 00:24:00,2011.0,1,1,0,24.0,24.9,13.3,48.0,SE,11,1007.5,0.0,2011-01-01 00:24:00
2011-01-01 00:34:00,2011.0,1,1,0,34.0,24.7,13.4,49.0,SE,11,1007.4,0.0,2011-01-01 00:34:00
2011-01-01 00:44:00,2011.0,1,1,0,44.0,24.1,13.3,51.0,ESE,9,1007.3,0.0,2011-01-01 00:44:00


In [354]:
all_df.drop(columns=['Year','Month','Day','Hour','Minute','Timestamp'], inplace=True)
all_df = all_df.drop_duplicates()   # Just repeating this so we are sure that there are no duplicates
all_df.head()

,Air Temp (degrees C),Dew Pt Temp (degrees C),Humidity (%),Wind Direction,Wind Speed (km/h),MSLP (hPa),Rainfall since 9 am (mm)
Timestamp,,,,,,,
2011-01-01 00:04:00,24.8,14.0,51.0,SE,11,1007.4,0.0
2011-01-01 00:14:00,24.8,13.3,48.0,SE,11,1007.5,0.0
2011-01-01 00:24:00,24.9,13.3,48.0,SE,11,1007.5,0.0
2011-01-01 00:34:00,24.7,13.4,49.0,SE,11,1007.4,0.0
2011-01-01 00:44:00,24.1,13.3,51.0,ESE,9,1007.3,0.0


### Check for Missing Values

In [355]:
# Make sure that all the data are in proper format/class as they should be

all_df['Air Temp (degrees C)'] = pd.to_numeric(all_df['Air Temp (degrees C)'], errors='coerce')
all_df['Dew Pt Temp (degrees C)'] = pd.to_numeric(all_df['Dew Pt Temp (degrees C)'], errors='coerce')
all_df['Humidity (%)'] = pd.to_numeric(all_df['Humidity (%)'], errors='coerce')
all_df['Wind Direction'] = all_df['Wind Direction'].astype(str)
all_df['Wind Direction'] = all_df['Wind Direction'].str.strip()
all_df['Wind Speed (km/h)'] = pd.to_numeric(all_df['Wind Speed (km/h)'], errors='coerce')
all_df['MSLP (hPa)'] = pd.to_numeric(all_df['MSLP (hPa)'], errors='coerce')
all_df['Rainfall since 9 am (mm)'] = pd.to_numeric(all_df['Rainfall since 9 am (mm)'], errors='coerce')
all_df.tail()

C:\Users\Zenbook\AppData\Local\Temp\ipykernel_1984\3404604326.py:1: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  all_df['Air Temp (degrees C)'] = pd.to_numeric(all_df['Air Temp (degrees C)'], errors='coerce')
C:\Users\Zenbook\AppData\Local\Temp\ipykernel_1984\3404604326.py:2: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  all_df['Dew Pt Temp (degrees C)'] = pd.to_numeric(all_df['Dew Pt Temp (degrees C)'], errors='coerce')
C:\Users\Zenbook\AppData\Local\Temp\ipykernel_1984\3404604326.py:3: SettingWithCopyWar

,Air Temp (degrees C),Dew Pt Temp (degrees C),Humidity (%),Wind Direction,Wind Speed (km/h),MSLP (hPa),Rainfall since 9 am (mm)
Timestamp,,,,,,,
2025-03-12 15:30:00,33.4,14.9,32.0,NW,20.0,1014.6,0.0
2025-03-12 15:34:00,33.5,14.5,31.0,NNW,26.0,1014.6,0.0
2025-03-12 16:00:00,34.4,14.8,30.0,WNW,26.0,1014.4,0.0
2025-03-12 16:30:00,33.6,14.1,30.0,WNW,26.0,1014.1,0.0
2025-03-12 16:30:00,33.6,14.1,30.0,NW,28.0,1014.1,0.0


In [356]:
all_df.isna().sum()

Air Temp (degrees C)          0
Dew Pt Temp (degrees C)       0
Humidity (%)                  0
Wind Direction                0
Wind Speed (km/h)             0
MSLP (hPa)                    0
Rainfall since 9 am (mm)    505
dtype: int64

In [357]:
all_df.describe()

,Air Temp (degrees C),Dew Pt Temp (degrees C),Humidity (%),Wind Speed (km/h),MSLP (hPa),Rainfall since 9 am (mm)
count,570653.000000,570653.000000,570653.000000,570653.000000,570653.000000,570148.000000
mean,15.912827,9.096631,66.798829,19.219853,1016.431970,0.792273
std,5.601476,4.063485,17.714121,34.157209,85.961523,2.979064
min,0.500000,-9.600000,0.000000,-9999.000000,-9999.000000,0.000000
25%,11.900000,6.100000,55.000000,11.000000,1012.200000,0.000000
50%,15.200000,8.700000,68.000000,17.000000,1017.300000,0.000000
75%,19.100000,11.800000,80.000000,26.000000,1022.400000,0.200000
max,43.500000,36.900000,100.000000,100.000000,1041.900000,54.600000


We can see that from **all_df.isna().sum()** only show missing values for Rainfall columns (505 missing values). 

But when we check **all_df.describe()**, there exists missing values in Wind Speed and MSLP as well but in the form of -9999.00 instead.

So, now we can try to deal with those missing values.

Since Wind Direction is a catogical data (not numeric), it doesn't show in this describe table. So, we should check for the missing value of Wind Direction seperately. We can do that listing out all the unique values in that Wind Direction column.

In [358]:
all_df['Wind Direction'].unique()

array(['SE', 'ESE', 'NE', 'NNE', 'ENE', 'W', 'N', 'SW', 'SSE', 'S', 'SSW',
       'WSW', 'NW', 'WNW', 'NNW', 'E', 'CALM', '1016.6', '1016.9',
       '1017.0', '1017.4', '1026.0', '1026.1', '1025.3', '1024.2',
       '1029.6', '1003.3', '1008.2', '1019.0', '1015.5', '1014.2',
       '1015.9', '1016.2', '1016.1', '1014.1', '1014.0', '1013.6',
       '1012.7', '1013.1', '1014.4', '1014.6', '1015.6', '1015.8',
       '1016.7', '1017.2', '1016.0', '1017.8', '1019.5', '1019.8',
       '1019.9', '1020.4', '1021.0', '1021.2', '1021.7', '1021.9',
       '1022.3', '1022.7', '1004.5', '1022.0', '1010.3', '1012.0',
       '1010.8', '-'], dtype=object)

And now we found another problem.....

In Wind Direction column, they should be no number!!!

In [359]:
# Go to one of the rows to see what's wrong with the data.
all_df[all_df['Wind Direction'].isin(['1017.0'])]

,Air Temp (degrees C),Dew Pt Temp (degrees C),Humidity (%),Wind Direction,Wind Speed (km/h),MSLP (hPa),Rainfall since 9 am (mm)
Timestamp,,,,,,,
2012-09-24 14:14:00,13.4,4.2,53.0,1017.0,0.0,0.0,0.0
2012-09-24 14:34:00,14.3,2.3,44.0,1017.0,0.0,0.0,0.0


From that table above we can see that the data for Wind Direction and Wind Speed are missing. The value 1017.0 that are currently in Wind Direction column are supposed to be in MSLP column. That means these data are shifted 2 columns to the left.

.....let's fix this.....(deep breath)....

### Dealing with shifted MSLP data

In [360]:
# Get all those rows that hold numeric data in Wind Direction column

valid_direction = ['N','NNE','NE','ENE','E','ESE','SE','SSE',
                   'S','SSW','SW','WSW','W','WNW','NW','NNW', 'CALM', '-']   # All the valid directions

# Create a df for not-shifted rows (Wind Direction column holds Valid Direction)
MSLP_not_shifted_df = all_df[all_df['Wind Direction'].isin(valid_direction)]

# Create a df for shifted rows (Wind Direction column DOES NOT hold valid direction)
MSLP_shifted_df = all_df[~all_df['Wind Direction'].isin(valid_direction)]

# Check if all the listed rows are actually shifted
MSLP_shifted_df

,Air Temp (degrees C),Dew Pt Temp (degrees C),Humidity (%),Wind Direction,Wind Speed (km/h),MSLP (hPa),Rainfall since 9 am (mm)
Timestamp,,,,,,,
2012-09-24 13:44:00,14.8,0.5,37.0,1016.6,0.0,0.0,0.0
2012-09-24 13:54:00,13.8,2.1,45.0,1016.9,0.0,0.0,0.0
2012-09-24 14:14:00,13.4,4.2,53.0,1017.0,0.0,0.0,0.0
2012-09-24 14:34:00,14.3,2.3,44.0,1017.0,0.0,0.0,0.0
2012-09-24 15:04:00,13.4,1.2,43.0,1017.4,0.0,0.0,0.0
2013-04-19 11:45:00,16.3,0.1,33.0,1026.0,0.0,0.0,0.0
2013-04-30 10:35:00,15.7,6.9,56.0,1026.1,0.0,0.0,0.0
2013-05-29 11:05:00,15.9,6.2,52.0,1025.3,0.0,0.0,0.0
2013-06-26 10:25:00,9.2,3.9,69.0,1024.2,0.0,0.0,0.0


In [361]:
# check if we had correctly capture NOT shifted rows
# There should be only VALID DIRECTION in Wind Direction column for this NOT shifted df

MSLP_not_shifted_df_try['Wind Direction'].unique()

array(['SE', 'ESE', 'NE', 'NNE', 'ENE', 'W', 'N', 'SW', 'SSE', 'S', 'SSW',
       'WSW', 'NW', 'WNW', 'NNW', 'E', 'CALM', '-'], dtype=object)

In [362]:
# Check if we had correctly capture the shifted rows
# There should be only numeric values in Wind Direction column for this shifted df

MSLP_shifted_df['Wind Direction'].unique()

array(['1016.6', '1016.9', '1017.0', '1017.4', '1026.0', '1026.1',
       '1025.3', '1024.2', '1029.6', '1003.3', '1008.2', '1019.0',
       '1015.5', '1014.2', '1015.9', '1016.2', '1016.1', '1014.1',
       '1014.0', '1013.6', '1012.7', '1013.1', '1014.4', '1014.6',
       '1015.6', '1015.8', '1016.7', '1017.2', '1016.0', '1017.8',
       '1019.5', '1019.8', '1019.9', '1020.4', '1021.0', '1021.2',
       '1021.7', '1021.9', '1022.3', '1022.7', '1004.5', '1022.0',
       '1010.3', '1012.0', '1010.8'], dtype=object)

In [363]:
# Rename the columns just so they match the actual data in the df
new_col=["Air Temp (degrees C)", "Dew Pt Temp (degrees C)", "Humidity (%)", "MSLP (hPa)", "Rainfall since 9 am (mm)", "Dummy1", "Dummy2"]

MSLP_shifted_df.columns = new_col
MSLP_shifted_df.head()

,Air Temp (degrees C),Dew Pt Temp (degrees C),Humidity (%),MSLP (hPa),Rainfall since 9 am (mm),Dummy1,Dummy2
Timestamp,,,,,,,
2012-09-24 13:44:00,14.8,0.5,37.0,1016.6,0.0,0.0,0.0
2012-09-24 13:54:00,13.8,2.1,45.0,1016.9,0.0,0.0,0.0
2012-09-24 14:14:00,13.4,4.2,53.0,1017.0,0.0,0.0,0.0
2012-09-24 14:34:00,14.3,2.3,44.0,1017.0,0.0,0.0,0.0
2012-09-24 15:04:00,13.4,1.2,43.0,1017.4,0.0,0.0,0.0


In [364]:
MSLP_shifted_df.drop(columns=['Dummy1','Dummy2'], inplace=True) # Dropping the 2 dummy columns as we don't need them

C:\Users\Zenbook\AppData\Local\Temp\ipykernel_1984\3692582120.py:1: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  MSLP_shifted_df.drop(columns=['Dummy1','Dummy2'], inplace=True) # Dropping the 2 dummy columns as we don't need them


In [365]:
# Create Wind Direction and Wind Speed column with '-' representing missing values
MSLP_shifted_df['Wind Direction'] = '-'
MSLP_shifted_df['Wind Speed (km/h)'] = -9999.00
MSLP_shifted_df.head()

C:\Users\Zenbook\AppData\Local\Temp\ipykernel_1984\1371292172.py:2: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  MSLP_shifted_df['Wind Direction'] = '-'
C:\Users\Zenbook\AppData\Local\Temp\ipykernel_1984\1371292172.py:3: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  MSLP_shifted_df['Wind Speed (km/h)'] = -9999.00


,Air Temp (degrees C),Dew Pt Temp (degrees C),Humidity (%),MSLP (hPa),Rainfall since 9 am (mm),Wind Direction,Wind Speed (km/h)
Timestamp,,,,,,,
2012-09-24 13:44:00,14.8,0.5,37.0,1016.6,0.0,-,-9999.0
2012-09-24 13:54:00,13.8,2.1,45.0,1016.9,0.0,-,-9999.0
2012-09-24 14:14:00,13.4,4.2,53.0,1017.0,0.0,-,-9999.0
2012-09-24 14:34:00,14.3,2.3,44.0,1017.0,0.0,-,-9999.0
2012-09-24 15:04:00,13.4,1.2,43.0,1017.4,0.0,-,-9999.0


In [371]:
all_df = pd.concat([MSLP_not_shifted_df, MSLP_shifted_df])  # Combining the non-shifted df with the shifted df (which has already been fixed)
all_df = all_df.sort_index()         # Sort all those rows back into place
all_df = all_df.drop_duplicates()    # Make sure that there are no duplicates
all_df

,Air Temp (degrees C),Dew Pt Temp (degrees C),Humidity (%),Wind Direction,Wind Speed (km/h),MSLP (hPa),Rainfall since 9 am (mm)
Timestamp,,,,,,,
2011-01-01 00:04:00,24.8,14.0,51.0,SE,11.0,1007.4,0.0
2011-01-01 00:14:00,24.8,13.3,48.0,SE,11.0,1007.5,0.0
2011-01-01 00:24:00,24.9,13.3,48.0,SE,11.0,1007.5,0.0
2011-01-01 00:34:00,24.7,13.4,49.0,SE,11.0,1007.4,0.0
2011-01-01 00:44:00,24.1,13.3,51.0,ESE,9.0,1007.3,0.0
...,...,...,...,...,...,...,...
2025-03-12 15:30:00,33.4,14.9,32.0,NW,20.0,1014.6,0.0
2025-03-12 15:34:00,33.5,14.5,31.0,NNW,26.0,1014.6,0.0
2025-03-12 16:00:00,34.4,14.8,30.0,WNW,26.0,1014.4,0.0


In [372]:
all_df['Wind Direction'].unique()

array(['SE', 'ESE', 'NE', 'NNE', 'ENE', 'W', 'N', 'SW', 'SSE', 'S', 'SSW',
       'WSW', 'NW', 'WNW', 'NNW', 'E', 'CALM', '-'], dtype=object)

Now, all the values in Wind Direction column are in proper format (valid direction).

But there are still missing values that we have to deal with.

### Dealing with Missing Values

In [373]:
# Missing values in Wind Speed and MSLP columns (in form of -9999.00)

cols = ['Wind Speed (km/h)', 'MSLP (hPa)']

all_df[all_df[cols].isin([-9999]).any(axis=1)]  # Printing this out just so we can observe how the -9999.00 data look like

,Air Temp (degrees C),Dew Pt Temp (degrees C),Humidity (%),Wind Direction,Wind Speed (km/h),MSLP (hPa),Rainfall since 9 am (mm)
Timestamp,,,,,,,
2012-09-24 13:44:00,14.8,0.5,37.0,-,-9999.0,1016.6,0.0
2012-09-24 13:54:00,13.8,2.1,45.0,-,-9999.0,1016.9,0.0
2012-09-24 14:14:00,13.4,4.2,53.0,-,-9999.0,1017.0,0.0
2012-09-24 14:34:00,14.3,2.3,44.0,-,-9999.0,1017.0,0.0
2012-09-24 15:04:00,13.4,1.2,43.0,-,-9999.0,1017.4,0.0
...,...,...,...,...,...,...,...
2024-04-15 10:30:00,17.8,10.4,62.0,W,9.0,-9999.0,0.0
2024-04-15 11:00:00,18.5,10.3,59.0,WNW,4.0,-9999.0,0.0
2024-10-10 20:00:00,13.5,5.9,60.0,S,15.0,-9999.0,0.0


In [374]:
# Missing values in Wind Direction columns (in form of '-')

all_df[all_df['Wind Direction'].isin(['-'])]

,Air Temp (degrees C),Dew Pt Temp (degrees C),Humidity (%),Wind Direction,Wind Speed (km/h),MSLP (hPa),Rainfall since 9 am (mm)
Timestamp,,,,,,,
2012-09-24 13:44:00,14.8,0.5,37.0,-,-9999.0,1016.6,0.0
2012-09-24 13:54:00,13.8,2.1,45.0,-,-9999.0,1016.9,0.0
2012-09-24 14:14:00,13.4,4.2,53.0,-,-9999.0,1017.0,0.0
2012-09-24 14:34:00,14.3,2.3,44.0,-,-9999.0,1017.0,0.0
2012-09-24 15:04:00,13.4,1.2,43.0,-,-9999.0,1017.4,0.0
...,...,...,...,...,...,...,...
2023-01-19 14:00:00,17.1,7.5,53.0,-,-9999.0,1017.2,0.0
2023-03-03 10:30:00,17.4,7.7,53.0,-,-9999.0,1019.2,0.0
2023-09-18 13:30:00,16.3,11.2,72.0,-,-9999.0,1014.4,0.0


At this point, the data are pretty much clear and in having the same format across all the rows.

Reminder, there are still missing values that we have to work on.

As mentioned above, 
- missing values of Wind Speed and MSLP are in the form of -9999.00
- missing values of Wind Direction are in the form of '-'

So, next step, we should try to deal with these missing vlues by dropping/filling/interpolating/predicting/etc.

I think for now we can simply download the 'all_df' as csv and then try to clean the data from that.

In [375]:
# all_df.to_csv('cleaned_Melbourne-1.csv')   # 